File: Day3/Problem_8.py

"""  Data Structure :- Arrays(1-D & 2-D Arrays)
Problem Name :- Merge Overlapping Intervals
Problem Statement :- Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals and return an array of the non-overlapping intervals that cover all the intervals in the input.

(Time Complexity :- O(n log n) and Space Complexity :- O(1))

"""

def merge_intervals(intervals):
    intervals.sort()  # Sort by start time

    merged = []

    for interval in intervals:
        if not merged or merged[-1][1] < interval[0]:
            merged.append(interval)
        else:
            merged[-1][1] = max(merged[-1][1], interval[1])

    return merged

if __name__ == "__main__":
    n = int(input("Enter number of intervals: "))
    intervals = []
    for _ in range(n):
        start, end = map(int, input().split())
        intervals.append([start, end])
    
    merged = merge_intervals(intervals)
    for interval in merged:
        print(interval[0], interval[1])

File: Day3/Problem_8.cpp
/* Data Structure :- Arrays(1-D & 2-D Arrays)
Problem Name :- Merge Overlapping Intervals
Problem Statement :- Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals and return an array of the non-overlapping intervals that cover all the intervals in the input.
*/

// TC : O(n log n) and SC : O(1)


#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> mergeIntervals(vector<vector<int>>& intervals, int n) {
    sort(intervals.begin(), intervals.end());

    vector<vector<int>> ans;

    for (int i = 0; i < n; i++) {
        if (ans.empty() || ans.back()[1] < intervals[i][0]) {
            ans.push_back(intervals[i]);
        } else {
            ans.back()[1] = max(ans.back()[1], intervals[i][1]);
        }
    }

    return ans;
}

int main() {
    int n;
    cin >> n;

    vector<vector<int>> intervals(n, vector<int>(2));

    for (int i = 0; i < n; i++) {
        cin >> intervals[i][0] >> intervals[i][1];
    }

    vector<vector<int>> merged = mergeIntervals(intervals, n);

    for (int i = 0; i < merged.size(); i++) {
        cout << merged[i][0] << " " << merged[i][1] << endl;
    }

    return 0;
}